File: modules/wordfold-translate/android/src/main/java/expo/modules/wordfoldtranslate/WordfoldTranslateModule.kt
package expo.modules.wordfoldtranslate

import com.google.mlkit.common.model.DownloadConditions
import com.google.mlkit.nl.translate.TranslateLanguage
import com.google.mlkit.nl.translate.Translation
import com.google.mlkit.nl.translate.TranslatorOptions
import expo.modules.kotlin.Promise
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

class WordfoldTranslateModule : Module() {
  override fun definition() = ModuleDefinition {
    Name("WordfoldTranslate")

    AsyncFunction("translate") { text: String, sourceCode: String, targetCode: String, promise: Promise ->
      val sourceLanguage = TranslateLanguage.fromLanguageTag(sourceCode)
      val targetLanguage = TranslateLanguage.fromLanguageTag(targetCode)
      if (sourceLanguage == null || targetLanguage == null) {
        promise.reject("E_LANGUAGE", "Unsupported translation language pair", null)
        return@AsyncFunction
      }

      val options = TranslatorOptions.Builder()
        .setSourceLanguage(sourceLanguage)
        .setTargetLanguage(targetLanguage)
        .build()
      val translator = Translation.getClient(options)
      val conditions = DownloadConditions.Builder().requireWifi().build()

      translator.downloadModelIfNeeded(conditions)
        .continueWithTask { translator.translate(text) }
        .addOnSuccessListener { result ->
          promise.resolve(result)
          translator.close()
        }
        .addOnFailureListener { error ->
          promise.reject("E_TRANSLATION", error.message ?: "Translation failed", error)
          translator.close()
        }
    }
  }
}

File: modules/wordfold-translate/ios/WordfoldTranslateModule.swift
import ExpoModulesCore
import MLKitTranslate

public class WordfoldTranslateModule: Module {
  public func definition() -> ModuleDefinition {
    Name("WordfoldTranslate")

    AsyncFunction("translate") { (text: String, sourceCode: String, targetCode: String, promise: Promise) in
      guard let sourceLanguage = self.language(for: sourceCode),
            let targetLanguage = self.language(for: targetCode) else {
        promise.reject("E_LANGUAGE", "Unsupported translation language pair")
        return
      }

      let options = TranslatorOptions(sourceLanguage: sourceLanguage, targetLanguage: targetLanguage)
      let translator = Translator.translator(options: options)
      let conditions = ModelDownloadConditions(allowsCellularAccess: false, allowsBackgroundDownloading: true)
      translator.downloadModelIfNeeded(with: conditions) { error in
        if let error {
          promise.reject("E_MODEL_DOWNLOAD", error.localizedDescription, error)
          return
        }
        translator.translate(text) { translatedText, error in
          if let error {
            promise.reject("E_TRANSLATION", error.localizedDescription, error)
          } else if let translatedText {
            promise.resolve(translatedText)
          } else {
            promise.reject("E_TRANSLATION", "Translation returned no text")
          }
        }
      }
    }
  }

  private func language(for code: String) -> TranslateLanguage? {
    switch code {
    case "en": return .english
    case "sk": return .slovak
    default: return nil
    }
  }
}
